File: frontend/src/types/money.ts
export type MoneyUserRole = 'admin' | 'developer' | 'viewer';

export interface MoneyCurrentUser {
  id: string;
  email: string;
  display_name: string;
  role: MoneyUserRole;
}

export interface AuthResponse {
  user: MoneyCurrentUser;
  access_token: string;
  refresh_token: string;
  expires_at: number;
}

export interface AuthMeResponse {
  user: MoneyCurrentUser;
}

export type MoneyFeatureType = 'area' | 'boulder' | 'problem' | 'trail' | 'topo' | 'poi' | 'drawing';
export type MoneyFeatureStatus = 'draft' | 'active' | 'archived' | 'scouted' | 'needs-work' | 'cleaning' | 'established' | 'project' | 'sent';
export type MoneyNoteVisibility = 'private' | 'team';
export type MoneyDevStatus = 'scouted' | 'needs-work' | 'cleaning' | 'established';
export type MoneyProblemStatus = 'project' | 'sent' | 'established';
export type MoneyNoteTargetType = 'project' | 'feature' | 'area' | 'boulder' | 'trail' | 'point' | 'none';
export type MoneyUploadBlockKind = 'photo' | 'sketch' | 'file' | 'topo';
export type MoneyArchiveMode = 'subtree' | 'promote_children';
export type MoneyTrailCategory = 'connector' | 'approach' | 'trail_to_area' | 'trail_to_destination';

export type MoneyPosition = [number, number];

export interface PointGeometry { type: 'Point'; coordinates: MoneyPosition; }
export interface LineStringGeometry { type: 'LineString'; coordinates: MoneyPosition[]; }
export interface PolygonGeometry { type: 'Polygon'; coordinates: MoneyPosition[][]; }
export interface FeatureGeometry { type: 'Feature'; geometry: MoneyGeometry; properties?: Record<string, unknown>; }
export interface FeatureCollectionGeometry { type: 'FeatureCollection'; features: FeatureGeometry[]; }

export type MoneyGeometry = PointGeometry | LineStringGeometry | PolygonGeometry;
export type MoneyGeoJSON = MoneyGeometry | FeatureGeometry | FeatureCollectionGeometry;

export interface MoneyProject {
  id: string;
  slug: string;
  name: string;
  center_lat: number;
  center_lon: number;
  default_zoom: number;
  created_at: string;
  updated_at: string;
}

export interface MoneyFeature {
  id: string;
  project_id: string;
  parent_feature_id?: string;
  feature_type: MoneyFeatureType;
  title: string;
  description?: string;
  status: MoneyFeatureStatus;
  geojson: MoneyGeoJSON;
  style: Record<string, unknown>;
  properties: Record<string, unknown>;
  min_lat?: number;
  min_lon?: number;
  max_lat?: number;
  max_lon?: number;
  sort_order?: number;
  external_ref?: string;
  import_source?: string;
  created_by: string;
  updated_by: string;
  created_at: string;
  updated_at: string;
}

export interface MoneyNoteBlock {
  kind: MoneyUploadBlockKind;
  upload_id?: string;
  url?: string;
  name?: string;
  metadata?: Record<string, unknown>;
}

export interface MoneyNote {
  id: string;
  project_id: string;
  feature_id?: string;
  target_type?: MoneyNoteTargetType;
  target_ref?: string;
  body: string;
  visibility: MoneyNoteVisibility;
  tags?: string[];
  blocks?: MoneyNoteBlock[];
  external_ref?: string;
  import_source?: string;
  created_by: string;
  updated_by: string;
  created_at: string;
  updated_at: string;
}

export interface MoneyUpload {
  id: string;
  project_id: string;
  feature_id?: string;
  note_id?: string;
  original_filename: string;
  content_type: string;
  byte_size: number;
  width?: number;
  height?: number;
  checksum_sha256: string;
  block_kind: MoneyUploadBlockKind;
  metadata: Record<string, unknown>;
  asset_kind: string;
  storage_backend: 'local' | 'r2';
  storage_bucket?: string;
  storage_region?: string;
  visibility: 'private';
  sync_status: 'available' | 'pending_upload' | 'deleted';
  uploaded_by: string;
  created_at: string;
}

export interface MoneyUploadDownloadURL {
  url: string;
  expires_at: string;
  proxy_url: string;
}

export interface MoneyPermissions { can_read: boolean; can_write: boolean; is_admin: boolean; }
export interface MoneyProjectResponse { project: MoneyProject; user: MoneyCurrentUser; permissions: MoneyPermissions; }
export interface MoneySnapshot { project: MoneyProject; features: MoneyFeature[]; note_counts: Record<string, number>; primary_uploads: Record<string, MoneyUpload>; }
export interface MoneyCragNode { feature: MoneyFeature; children?: MoneyCragNode[] | null; boulders?: MoneyCragNode[] | null; problems?: MoneyCragNode[] | null; }
export interface MoneyCragSnapshot { project: MoneyProject; root: MoneyCragNode | null; trails?: MoneyCragNode[] | null; notes?: MoneyNote[] | null; uploads?: MoneyUpload[] | null; }
export interface MoneyFeatureDetail { feature: MoneyFeature; notes: MoneyNote[] | null; uploads: MoneyUpload[] | null; }
export interface MoneyTrashItem { id: string; title: string; feature_type: MoneyFeatureType; parent_feature_id?: string; path: string[]; deleted_at: string; updated_at: string; descendant_count: number; }
export interface MoneyTrashResponse { items: MoneyTrashItem[]; }

export interface MoneyFeatureRequest {
  parent_feature_id?: string | null;
  feature_type: MoneyFeatureType;
  title: string;
  description?: string | null;
  status: MoneyFeatureStatus;
  geojson: MoneyGeoJSON;
  style: Record<string, unknown>;
  properties: Record<string, unknown>;
  sort_order?: number;
}

export interface MoneyAreaRequest { parent_feature_id?: string | null; title: string; description?: string | null; geojson: MoneyGeoJSON; properties: Record<string, unknown>; }
export interface MoneyAreaGeometryRequest { geojson: MoneyGeoJSON; }
export interface MoneyMoveFeatureRequest { parent_feature_id: string | null; sort_order?: number; }
export interface MoneyBoulderRequest { parent_feature_id: string; title: string; description?: string | null; dev_status: MoneyDevStatus; geojson: MoneyGeoJSON; properties: Record<string, unknown>; }
export interface MoneyProblemRequest { boulder_id: string; name: string; grade: string; status: MoneyProblemStatus; stars: number; fa?: string | null; types: string[]; description?: string | null; properties?: Record<string, unknown>; }
export interface MoneyBoulderStatusRequest { dev_status: MoneyDevStatus; }

export interface MoneyNoteRequest {
  body: string;
  visibility: MoneyNoteVisibility;
  target_type?: MoneyNoteTargetType;
  target_ref?: string | null;
  tags?: string[];
  blocks?: MoneyNoteBlock[];
}

export interface MoneyFeatureFilters { type?: MoneyFeatureType | 'all'; status?: MoneyFeatureStatus | 'all'; search?: string; }
export interface MoneyBBox { minLon: number; minLat: number; maxLon: number; maxLat: number; }

File: backend/internal/models/money.go
package models

import (
	"encoding/json"
	"time"
)

const (
	MoneyFeatureArea    = "area"
	MoneyFeatureBoulder = "boulder"
	MoneyFeatureProblem = "problem"
	MoneyFeatureTrail   = "trail"
	MoneyFeatureTopo    = "topo"
	MoneyFeaturePOI     = "poi"
	MoneyFeatureDrawing = "drawing"

	MoneyStatusDraft       = "draft"
	MoneyStatusActive      = "active"
	MoneyStatusArchived    = "archived"
	MoneyStatusScouted     = "scouted"
	MoneyStatusNeedsWork   = "needs-work"
	MoneyStatusCleaning    = "cleaning"
	MoneyStatusEstablished = "established"
	MoneyStatusProject     = "project"
	MoneyStatusSent        = "sent"

	MoneyNotePrivate = "private"
	MoneyNoteTeam    = "team"

	MoneyTrailCategoryConnector          = "connector"
	MoneyTrailCategoryApproach           = "approach"
	MoneyTrailCategoryTrailToArea        = "trail_to_area"
	MoneyTrailCategoryTrailToDestination = "trail_to_destination"
)

type MoneyProject struct {
	ID          string    `json:"id"`
	Slug        string    `json:"slug"`
	Name        string    `json:"name"`
	CenterLat   float64   `json:"center_lat"`
	CenterLon   float64   `json:"center_lon"`
	DefaultZoom float64   `json:"default_zoom"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type MoneyFeature struct {
	ID              string          `json:"id"`
	ProjectID       string          `json:"project_id"`
	ParentFeatureID *string         `json:"parent_feature_id,omitempty"`
	FeatureType     string          `json:"feature_type"`
	Title           string          `json:"title"`
	Description     *string         `json:"description,omitempty"`
	Status          string          `json:"status"`
	GeoJSON         json.RawMessage `json:"geojson"`
	Style           json.RawMessage `json:"style"`
	Properties      json.RawMessage `json:"properties"`
	MinLat          *float64        `json:"min_lat,omitempty"`
	MinLon          *float64        `json:"min_lon,omitempty"`
	MaxLat          *float64        `json:"max_lat,omitempty"`
	MaxLon          *float64        `json:"max_lon,omitempty"`
	SortOrder       int             `json:"sort_order"`
	ExternalRef     *string         `json:"external_ref,omitempty"`
	ImportSource    *string         `json:"import_source,omitempty"`
	CreatedBy       string          `json:"created_by"`
	UpdatedBy       string          `json:"updated_by"`
	CreatedAt       time.Time       `json:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at"`
}

type MoneyNoteBlock struct {
	Kind     string          `json:"kind"`
	UploadID *string         `json:"upload_id,omitempty"`
	URL      *string         `json:"url,omitempty"`
	Name     *string         `json:"name,omitempty"`
	Metadata json.RawMessage `json:"metadata,omitempty"`
}

type MoneyNote struct {
	ID           string          `json:"id"`
	ProjectID    string          `json:"project_id"`
	FeatureID    *string         `json:"feature_id,omitempty"`
	TargetType   string          `json:"target_type"`
	TargetRef    *string         `json:"target_ref,omitempty"`
	Body         string          `json:"body"`
	Visibility   string          `json:"visibility"`
	Tags         []string        `json:"tags"`
	Blocks       json.RawMessage `json:"blocks"`
	ExternalRef  *string         `json:"external_ref,omitempty"`
	ImportSource *string         `json:"import_source,omitempty"`
	CreatedBy    string          `json:"created_by"`
	UpdatedBy    string          `json:"updated_by"`
	CreatedAt    time.Time       `json:"created_at"`
	UpdatedAt    time.Time       `json:"updated_at"`
}

type MoneyUpload struct {
	ID                  string          `json:"id"`
	ProjectID           string          `json:"project_id"`
	FeatureID           *string         `json:"feature_id,omitempty"`
	NoteID              *string         `json:"note_id,omitempty"`
	OriginalFilename    string          `json:"original_filename"`
	Title               *string         `json:"title,omitempty"`
	Comments            *string         `json:"comments,omitempty"`
	StorageKey          string          `json:"-"`
	ContentType         string          `json:"content_type"`
	ByteSize            int64           `json:"byte_size"`
	Width               *int            `json:"width,omitempty"`
	Height              *int            `json:"height,omitempty"`
	ChecksumSHA256      string          `json:"checksum_sha256"`
	BlockKind           string          `json:"block_kind"`
	Metadata            json.RawMessage `json:"metadata"`
	AssetKind           string          `json:"asset_kind"`
	StorageBackend      string          `json:"storage_backend"`
	StorageBucket       *string         `json:"storage_bucket,omitempty"`
	StorageRegion       *string         `json:"storage_region,omitempty"`
	StorageETag         *string         `json:"storage_etag,omitempty"`
	StorageVersionID    *string         `json:"storage_version_id,omitempty"`
	Visibility          string          `json:"visibility"`
	SyncStatus          string          `json:"sync_status"`
	DeletedAt           *time.Time      `json:"deleted_at,omitempty"`
	DeletedBy           *string         `json:"deleted_by,omitempty"`
	DeleteRequestedAt   *time.Time      `json:"delete_requested_at,omitempty"`
	PhysicallyDeletedAt *time.Time      `json:"physically_deleted_at,omitempty"`
	UploadedBy          string          `json:"uploaded_by"`
	CreatedAt           time.Time       `json:"created_at"`
	UpdatedAt           time.Time       `json:"updated_at"`
}

type MoneyUploadDownloadURL struct {
	URL       string    `json:"url"`
	ExpiresAt time.Time `json:"expires_at"`
	ProxyURL  string    `json:"proxy_url"`
}

type MoneyPermissions struct {
	CanRead  bool `json:"can_read"`
	CanWrite bool `json:"can_write"`
	IsAdmin  bool `json:"is_admin"`
}

type MoneyProjectResponse struct {
	Project     MoneyProject     `json:"project"`
	User        CurrentUser      `json:"user"`
	Permissions MoneyPermissions `json:"permissions"`
}

type MoneySnapshot struct {
	Project        MoneyProject           `json:"project"`
	Features       []MoneyFeature         `json:"features"`
	NoteCounts     map[string]int         `json:"note_counts"`
	PrimaryUploads map[string]MoneyUpload `json:"primary_uploads"`
}

type MoneyCragSnapshot struct {
	Project MoneyProject    `json:"project"`
	Root    *MoneyCragNode  `json:"root"`
	Trails  []MoneyCragNode `json:"trails"`
	Notes   []MoneyNote     `json:"notes"`
	Uploads []MoneyUpload   `json:"uploads"`
}

type MoneyCragNode struct {
	Feature  MoneyFeature    `json:"feature"`
	Children []MoneyCragNode `json:"children"`
	Boulders []MoneyCragNode `json:"boulders"`
	Problems []MoneyCragNode `json:"problems"`
}

type MoneyFeatureDetail struct {
	Feature MoneyFeature  `json:"feature"`
	Notes   []MoneyNote   `json:"notes"`
	Uploads []MoneyUpload `json:"uploads"`
}

type MoneyTrashItem struct {
	ID              string    `json:"id"`
	Title           string    `json:"title"`
	FeatureType     string    `json:"feature_type"`
	ParentFeatureID *string   `json:"parent_feature_id,omitempty"`
	Path            []string  `json:"path"`
	DeletedAt       time.Time `json:"deleted_at"`
	UpdatedAt       time.Time `json:"updated_at"`
	DescendantCount int       `json:"descendant_count"`
}

type MoneyTrashResponse struct {
	Items []MoneyTrashItem `json:"items"`
}

type MoneyFeatureRequest struct {
	ParentFeatureID *string         `json:"parent_feature_id,omitempty"`
	FeatureType     string          `json:"feature_type"`
	Title           string          `json:"title"`
	Description     *string         `json:"description"`
	Status          string          `json:"status"`
	GeoJSON         json.RawMessage `json:"geojson"`
	Style           json.RawMessage `json:"style"`
	Properties      json.RawMessage `json:"properties"`
	SortOrder       int             `json:"sort_order"`
	ExternalRef     *string         `json:"external_ref,omitempty"`
	ImportSource    *string         `json:"import_source,omitempty"`
}

type MoneyCragAreaRequest struct {
	ParentFeatureID *string         `json:"parent_feature_id,omitempty"`
	Title           string          `json:"title"`
	Description     *string         `json:"description"`
	GeoJSON         json.RawMessage `json:"geojson"`
	Properties      json.RawMessage `json:"properties"`
}

type MoneyAreaGeometryRequest struct {
	GeoJSON json.RawMessage `json:"geojson"`
}

type MoneyArchiveMode string

const (
	MoneyArchiveModeSubtree         MoneyArchiveMode = "subtree"
	MoneyArchiveModePromoteChildren MoneyArchiveMode = "promote_children"
)

type MoneyArchiveFeatureRequest struct {
	Mode MoneyArchiveMode `json:"mode"`
}

type MoneyMoveFeatureRequest struct {
	ParentFeatureID *string `json:"parent_feature_id"`
	SortOrder       *int    `json:"sort_order,omitempty"`
}

type MoneyCragBoulderRequest struct {
	ParentFeatureID string          `json:"parent_feature_id"`
	Title           string          `json:"title"`
	Description     *string         `json:"description"`
	DevStatus       string          `json:"dev_status"`
	GeoJSON         json.RawMessage `json:"geojson"`
	Properties      json.RawMessage `json:"properties"`
}

type MoneyCragProblemRequest struct {
	BoulderID   string          `json:"boulder_id"`
	Name        string          `json:"name"`
	Grade       string          `json:"grade"`
	Status      string          `json:"status"`
	Stars       int             `json:"stars"`
	FA          *string         `json:"fa,omitempty"`
	Types       []string        `json:"types"`
	Description *string         `json:"description"`
	Properties  json.RawMessage `json:"properties"`
}

type MoneyBoulderStatusRequest struct {
	DevStatus string `json:"dev_status"`
}

type MoneyNoteRequest struct {
	Body       string          `json:"body"`
	Visibility string          `json:"visibility"`
	TargetType string          `json:"target_type"`
	TargetRef  *string         `json:"target_ref,omitempty"`
	Tags       []string        `json:"tags"`
	Blocks     json.RawMessage `json:"blocks"`
}

type MoneyUploadMetadataRequest struct {
	Title    *string `json:"title"`
	Comments *string `json:"comments"`
}

type BBox struct {
	MinLon float64
	MinLat float64
	MaxLon float64
	MaxLat float64
}

type MoneyFeatureFilter struct {
	FeatureType     string
	Status          string
	BBox            *BBox
	UpdatedAfter    *time.Time
	IncludeArchived bool
}
